src/handlers/expiry/providers/TestDataProvider.ts

```diff
@@ -0,0 +1,90 @@
+import moment from "moment";
+import { HTTPError } from "../../../models/HTTPError";
+import {
+  TEST_STATUS,
+  MESSAGES,
+  TEST_CODES_FOR_CALCULATING_EXPIRY
+} from "../../../assets/Enums";
+
+import { ITestResult } from "../../../models/ITestResult";
+import { TestResultsDAO } from "../../../models/TestResultsDAO";
+import { DateProvider } from "./DateProvider";
+import { ITestDataProvider } from "./ITestDataProvider";
+import { Service } from "../../../models/injector/ServiceDecorator";
+
+// TODO: Add unit tests data provider
+// 1- missing attributes scenario
+// 2- postivie scenario
+
+@Service()
+export class TestDataProvider implements ITestDataProvider {
+    public testResultsDAO: TestResultsDAO | undefined;
+
+  // getTestResult
+// Do we use existing functionality or write new function
+  public async getTestHistory(systemNumber: string): Promise<ITestResult[]> {
+    // testService object
+    const fromDateTime = new Date(1970, 1, 1);
+    const toDateTime = new Date();
+    let result: ITestResult[] = [];
+    try {
+      this.testResultsDAO = this.testResultsDAO as TestResultsDAO;
+      const data = await this.testResultsDAO.getBySystemNumber(systemNumber);
+      if (data && data.Count) {
+        result = data.Items as ITestResult[];
+        result = result.filter(
+          (testResult: {
+            testStartTimestamp: string | number | Date;
+            testEndTimestamp: string | number | Date;
+          }) => {
+            return (
+              moment(testResult.testStartTimestamp).isAfter(fromDateTime) &&
+              moment(testResult.testEndTimestamp).isBefore(toDateTime)
+            );
+          }
+        );
+        result = result.filter(
+          (test) => test.testStatus === TEST_STATUS.SUBMITTED
+        );
+      }
+      return result;
+    } catch (error) {
+      if (!(error instanceof HTTPError)) {
+        console.log(error);
+        error = new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR);
+      }
+      throw error;
+    }
+  }
+
+  public async getMostRecentExpiryDate(systemNumber: string): Promise<Date> {
+    let maxDate = new Date(1970, 1, 1);
+    const testResults = await this.getTestHistory(systemNumber);
```

Comment:
Throwing error makes more sense in my opinion than to return a default date. This would make sure that we don't create an expiry using a wrong value for recent expiry. @alex-deloitte  @JonLazariniDD  I would like your feedback on this. Does Dynamo throw an error if no result is found?

Reply:
I didn't test this behaviour in Dynamodb but I'd rather not return a default date. We will not need another method to check if the testHistory returned a valid date or not

Reply:
This is a good point and thanks for pointing this out, I think it is important to avoid mutating the original input where possible so we can 'guarantee' that possible side effects.
If my understanding is no wrong we set the date in the date provider which is great and act as the main conduit for it.
I think we could set up a session with the business to agree on how we should handle that, in theory if the Data layer fails we could we still might want to know what should return as an error. Apologies on not helping too much on this but I think it is more checking with architects and the business  requirements we currently have.